### ISSUE TITLE
Slang.D backward mode doesn't run lines of code after a for loop

### ISSUE THREAD

**Issue by yashbelhe** [shader-slang/slang#6394]
The code below shows a simple example of the bug. 

```
[Differentiable]
[AutoPyBindCUDA]
[CUDAKernel]
void run(
    DiffTensorView<float> x, DiffTensorView<float> ret)
{
    int idx = (cudaBlockIdx() * cudaBlockDim() + cudaThreadIdx()).x;
    if (idx >= 1) return;

    if (idx == 0) {
        printf("entered the kernel \n");
    }

    for (int i = 0; i < 1; i++) {
        if (idx > 0) {
            printf("this doesn't print since idx>0 is never true \n");
            return;
        }

        if (idx == 0) {
            printf("inside the loop \n");
        }
    }

    if (idx == 0) {
        printf("reached outside \n");
    }
    ret[idx] = x[idx,0];
}
```

In forward mode, when the code is run with x=[0.1,0.2], the following is printed
```
entered the kernel 
inside the loop 
reached outside 
entered the kernel 
inside the loop 
entered the kernel 
inside the loop
```
and x.grad = [0.0, 0.0], which is incorrect.

Ideally, the code should produce
```
entered the kernel 
inside the loop 
reached outside 
entered the kernel 
inside the loop 
reached outside 
entered the kernel 
inside the loop 
reached outside
```
and x.grad=[1,0].

A slightly modified version of the code, which comments out the for loop doesn't produce this bug

```
[Differentiable]
[AutoPyBindCUDA]
[CUDAKernel]
void run(
    DiffTensorView<float> x, DiffTensorView<float> ret)
{
    int idx = (cudaBlockIdx() * cudaBlockDim() + cudaThreadIdx()).x;
    if (idx >= 1) return;

    if (idx == 0) {
        printf("entered the kernel \n");
    }

    // for (int i = 0; i < 1; i++) {
        if (idx > 0) {
            printf("this doesn't print since idx>0 is never true \n");
            return;
        }

        if (idx == 0) {
            printf("inside the loop \n");
        }
    // }

    if (idx == 0) {
        printf("reached outside \n");
    }
    ret[idx] = x[idx,0];
}
```
The snippet above produces the right output with the correct print statements and gradients. This means the for loop is correlated with the bug.

Additionally the bug seems to correlated with the return statement within the for loop. The first if statement within the for loop never evaluates to true, yet, its presence causes the bug.

If you remove that return statement, we get the following code, 
```
[Differentiable]
[AutoPyBindCUDA]
[CUDAKernel]
void run(
    DiffTensorView<float> x, DiffTensorView<float> ret)
{
    int idx = (cudaBlockIdx() * cudaBlockDim() + cudaThreadIdx()).x;
    if (idx >= 1) return;

    if (idx == 0) {
        printf("entered the kernel \n");
    }

    for (int i = 0; i < 1; i++) {
        if (idx > 0) {
            printf("this doesn't print since idx>0 is never true \n");
            // return;
        }

        if (idx == 0) {
            printf("inside the loop \n");
        }
    }

    if (idx == 0) {
        printf("reached outside \n");
    }
    ret[idx] = x[idx,0];
}
```
which produces the correct output.

The following python code can be used to invoke the slang code above (it assumes the slang code is in bug.slang):
```
# %%
import torch
import slangtorch

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
torch.set_default_device(DEVICE)

class SlangShader(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, shader):
        ret = torch.zeros(x.shape[0], requires_grad=True)
        shader.run(x=x, ret=ret).launchRaw(
            blockSize=(32, 1, 1),
            gridSize=(x.shape[0] // 32 + 1, 1, 1)
        )
        ctx.save_for_backward(x, ret)
        ctx.shader = shader
        return ret

    @staticmethod
    def backward(ctx, grad_output):
        x, ret = ctx.saved_tensors
        d_x = torch.zeros_like(x)
        ctx.shader.run.bwd(
            x=(x, d_x), ret=(ret, grad_output.contiguous())
        ).launchRaw(
            blockSize=(32, 1, 1),
            gridSize=(x.shape[0] // 32 + 1, 1, 1)
        )
        return d_x, None

class BugReport(torch.nn.Module):
    def __init__(self, shader_file):
        super().__init__()
        self.shader = slangtorch.loadModule(shader_file)
        
    def forward(self, x):
        return SlangShader.apply(x, self.shader)

# Test
x = torch.tensor([0.1, 0.2])
x.requires_grad = True

shader = BugReport("bug.slang")
ret = shader(x)
ret.sum().backward()

print(x.grad)
```

Thanks for the help in advance!
